### ISSUE TITLE
更新到 linux kernel 5.17.0.r7

### ISSUE THREAD

**Issue by cpkt9762** [memflow/memflow-kvm#7]

**Comment by cpkt9762** [comment]
`static int get_sorted_memslots(struct kvm_memslots *slots, int max_slots, vm_memslot_t *slots_out)
{
	struct kvm_memory_slot *slot;
	int slot_count, i, bkt=0;

	int idx = slots->node_idx;

	slot_count = 0;  

	kvm_for_each_memslot(slot, bkt, slots) { 

        slots_out[slot_count++] = (vm_memslot_t) {
				.base = gfn_to_gpa(slot->base_gfn),
				.host_base = slot->userspace_addr,
				.map_size = gfn_to_gpa(slot->npages)
			};

	  };
 
	sort(slots_out, slot_count, sizeof(*slots_out), memslot_compare, NULL);

	//TODO: coalesce nearby slots

	return slot_count;
}
static int get_vm_info(struct kvm *kvm, vm_info_t __user *user_info)
{
/**/
	if (kernel_info.slot_count > HASH_SIZE(kvm_memslots(kvm)->id_hash))
		kernel_info.slot_count = HASH_SIZE(kvm_memslots(kvm)->id_hash);

/**/
}
`

**Comment by h33p** [comment]
Being fluent in many different languages, I think I can understand what you want to say.

Ideally, `get_vm_info` would need to use `kvm_for_each_memslot` to count the number of slots, it would also need to be gated to take that path on 5.17 upwards. 

`get_sorted_memslots` change seems way better than what I had, however, you need to check against `max_slots` to not overflow the buffer. Another check was removed, for `npages` being not 0. I'm not sure exactly if it is needed, but I believe I had it in for a reason. So that should re-appear in the loop.

I am rather busy at the moment, but I would be glad to accept a PR with the changes.

**Comment by cpkt9762** [comment]
我使用这个kvm_for_each_memslot的原因是因为"struct kvm_memslots "在 linux kernel 5.17.0 中发生啦变化,是下面这个样子
```
struct kvm_memslots {
	u64 generation;
	/* The mapping table from slot id to the index in memslots[]. */
	short id_to_index[KVM_MEM_SLOTS_NUM];
	atomic_t last_used_slot;
	int used_slots;
	struct kvm_memory_slot memslots[];
};
```
kvm_for_each_memslot 在新的内核5.17.0中"kvm_host.h:836 line"声明为
#define kvm_for_each_memslot(memslot, slots)				\
	for (memslot = &slots->memslots[0];				\
	     memslot < slots->memslots + slots->used_slots; memslot++)	\
		if (WARN_ON_ONCE(!memslot->npages)) {			\
		} else

max_slots 我忘记做啦检查,谢谢你的提示,我后面自己加上

**Comment by weewoo22** [comment]
This is resolved by 61c8bdb097ce629bfdafdb0b2f6173a521756712. Tested on `5.17.1`:
```command
$ uname -r
5.17.1
$ lsmod | rg memflow
memflow                20480  0
kvm                  1048576  2 memflow,kvm_intel
```
